File: scripts/family_update.py
"""Test candidate family locks in disposable standalone CI checkouts."""
import copy
import json
import os
from pathlib import Path
import re
import tempfile

from family import Family, FamilyError, git, run, write_atomic


def api(endpoint):
    return json.loads(run("gh", "api", endpoint, capture=True))


def successful(repo, sha):
    checks = api(f"repos/{repo['slug']}/commits/{sha}/check-runs?filter=latest&per_page=100")
    return any(c["name"] == repo["required_check"] and c["head_sha"] == sha
               and c["status"] == "completed" and c["conclusion"] == "success"
               and c["app"]["slug"] == "github-actions" for c in checks["check_runs"])


def eligible(family, repo):
    path = family.path(repo)
    branch = repo["default_branch"]
    git(path, "fetch", "--no-tags", repo["github"],
        f"refs/heads/{branch}:refs/remotes/origin/{branch}")
    git(path, "fetch", "--no-tags", repo["github"], "refs/tags/ci-*:refs/tags/ci-*")
    # An immutable CI tag is eligible only while reachable from the default branch
    # and backed by the aggregate GitHub Actions check for that exact commit.
    tags = set(git(path, "tag", "--list", "ci-*", capture=True).splitlines())
    for sha in git(path, "rev-list", f"refs/remotes/origin/{branch}", capture=True).splitlines():
        if f"ci-{sha}" in tags:
            tagged = git(path, "rev-parse", f"refs/tags/ci-{sha}^{{commit}}", capture=True)
            if tagged == sha and successful(repo, sha):
                return sha
    return family.pins[repo["name"]]["commit"]


def lock_text(original, pins):
    def replace_block(match):
        block = match.group(0)
        name = re.search(r'^repo = "([^"]+)"$', block, re.M).group(1)
        pin = pins[name]
        block = re.sub(r'^tag = "[^"]+"$', f'tag = "{pin["tag"]}"', block, flags=re.M)
        return re.sub(r'^commit = "[^"]+"$', f'commit = "{pin["commit"]}"', block, flags=re.M)
    return re.sub(r'\[\[repo\]\][\s\S]*?(?=\[\[repo\]\]|\Z)', replace_block, original)


def update(family):
    family.clean(family.hub)
    for repo in family.components():
        if family.existing(repo):
            family.clean(family.path(repo))
    family.bootstrap()
    pins = copy.deepcopy(family.pins)
    before = {p: (family.hub / p).read_text() for p in ("family.lock", "Cargo.lock")}
    for repo in family.components():
        sha = eligible(family, repo)
        path = family.path(repo)
        head = git(path, "rev-parse", "HEAD", capture=True)
        if git(path, "merge-base", "--is-ancestor", head, sha, check=False).returncode:
            raise FamilyError(f"{repo['name']}: candidate would leave divergent local work behind")
        if sha != pins[repo["name"]]["commit"]:
            pins[repo["name"]].update(commit=sha, tag=f"ci-{sha}")
    candidate = lock_text(before["family.lock"], pins)
    if candidate == before["family.lock"]:
        print("family pull: no newer successful component revisions")
        return
    target = family.hub / "target"
    target.mkdir(exist_ok=True)
    # Exact-SHA integration uses an automatically removed standalone checkout.
    # It never registers a Git worktree or moves a developer's component heads.
    with tempfile.TemporaryDirectory(prefix="family-ci-", dir=target) as temporary:
        hub = Path(temporary) / "jankurai"
        run("git", "clone", "--no-hardlinks", "--no-checkout", family.hub, hub)
        git(hub, "checkout", "--detach", git(family.hub, "rev-parse", "HEAD", capture=True))
        write_atomic(hub / "family.lock", candidate)
        sandbox = Family(hub)
        sandbox.bootstrap(restore=True)
        sandbox.fuse(copy_lock=False)
        # Strip publication credentials before running component build/test code.
        env = os.environ.copy()
        for key in ("GH_TOKEN", "GITHUB_TOKEN", "FAMILY_AUTOMATION_TOKEN"):
            env.pop(key, None)
        run("cargo", "generate-lockfile", cwd=sandbox.fusion, env=env)
        cargo = (sandbox.fusion / "Cargo.lock").read_text()
        write_atomic(hub / "Cargo.lock", cargo)
        run("bash", "scripts/family.sh", "check", cwd=hub, env=env)
        for name, content in before.items():
            if (family.hub / name).read_text() != content:
                raise FamilyError(f"{name} changed while testing candidate; accepted locks preserved")
        for repo in family.components():
            family.clean(family.path(repo))
        # No write to accepted locks occurs until all candidate checks pass.
        write_atomic(family.hub / "Cargo.lock", cargo)
        write_atomic(family.hub / "family.lock", candidate)
    print("family pull: validated candidate locks ready for a protected pull request")

File: scripts/family-update.mjs
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { atomicWrite, buildEnvironment, clean, git, gitText, run, temporaryCI } from './family-lib.mjs';
import {
  beginPairedJournal, captureFileIdentity, captureSource, commitPairedReplace,
  identityMeta, operation, recordAfterImages, setJournalState, assertIdentity,
} from './family-operation.mjs';

export function api(endpoint, body, method) {
  const args = ['gh', 'api', endpoint];
  if (method) args.push('--method', method);
  if (body !== undefined) args.push('--input', '-');
  const output = run(args, { capture: true, input: body === undefined ? undefined : JSON.stringify(body) });
  return output ? JSON.parse(output) : undefined;
}
export function successful(repo, sha, request = api) {
  const checks = request(`repos/${repo.slug}/commits/${sha}/check-runs?filter=latest&per_page=100`).check_runs;
  return checks.some(check => check.name === repo.required_check && check.head_sha === sha &&
    check.status === 'completed' && check.conclusion === 'success' && check.app.slug === 'github-actions');
}
export function eligible(family, repo) {
  const directory = family.path(repo), branch = repo.default_branch;
  const env = buildEnvironment();
  git(directory, ['fetch', '--no-tags', repo.github, `refs/heads/${branch}:refs/remotes/origin/${branch}`], { env });
  git(directory, ['fetch', '--no-tags', repo.github, 'refs/tags/ci-*:refs/tags/ci-*'], { env });
  const tags = new Set(gitText(directory, 'tag', '--list', 'ci-*').split('\n'));
  for (const sha of gitText(directory, 'rev-list', `refs/remotes/origin/${branch}`).split('\n')) {
    if (tags.has(`ci-${sha}`) && gitText(directory, 'rev-parse', `refs/tags/ci-${sha}^{commit}`) === sha && successful(repo, sha)) return sha;
  }
  return family.pins.get(repo.name).commit;
}
export function lockText(original, pins) {
  return original.replace(/\[\[repo\]\][\s\S]*?(?=\[\[repo\]\]|$)/g, block => {
    const name = /^repo = "([^"]+)"$/m.exec(block)?.[1];
    if (!name) throw new Error('invalid lock block');
    const pin = pins.get(name);
    return block.replace(/^tag = "[^"]+"$/m, `tag = "${pin.tag}"`).replace(/^commit = "[^"]+"$/m, `commit = "${pin.commit}"`);
  });
}

function captureSources(family) {
  const components = {};
  for (const repo of family.components()) {
    if (!family.existing(repo)) continue;
    components[repo.name] = captureSource(family.path(repo));
  }
  return { hub: captureSource(family.hub), components };
}

function assertSources(family, expected) {
  const hub = captureSource(family.hub);
  if (hub.head !== expected.hub.head || hub.tree !== expected.hub.tree) {
    throw new Error('hub HEAD/tree changed while validating; accepted locks preserved');
  }
  for (const repo of family.components()) {
    if (!family.existing(repo)) continue;
    const current = captureSource(family.path(repo));
    const prior = expected.components[repo.name];
    if (!prior || current.head !== prior.head || current.tree !== prior.tree) {
      throw new Error(`${repo.name}: HEAD/tree changed while validating; accepted locks preserved`);
    }
  }
}

function updateWithTx(family, hooks, tx) {
  clean(family.hub);
  for (const repo of family.components()) if (family.existing(repo)) clean(family.path(repo));
  family.bootstrap();
  const source = captureSources(family);
  const beforeIdentities = Object.fromEntries(
    ['family.lock', 'Cargo.lock'].map(name => [name, captureFileIdentity(path.join(family.hub, name))]),
  );
  beginPairedJournal(tx, {
    source,
    locks: beforeIdentities,
  });
  const pins = new Map([...family.pins].map(([name, pin]) => [name, { ...pin }]));
  for (const repo of family.components()) {
    const sha = (hooks.eligible ?? eligible)(family, repo), directory = family.path(repo);
    if (git(directory, ['merge-base', '--is-ancestor', gitText(directory, 'rev-parse', 'HEAD'), sha], { check: false }).status !== 0) {
      throw new Error(`${repo.name}: candidate would leave divergent local work behind`);
    }
    if (sha !== pins.get(repo.name).commit) Object.assign(pins.get(repo.name), { commit: sha, tag: `ci-${sha}` });
  }
  const candidate = lockText(beforeIdentities['family.lock'].bytes.toString('utf8'), pins);
  if (candidate === beforeIdentities['family.lock'].bytes.toString('utf8')) {
    console.log('family pull: no newer successful revisions');
    setJournalState(tx, 'rolled-back');
    return;
  }
  setJournalState(tx, 'validating');
  temporaryCI(path.join(family.hub, 'target'), 'family-ci-', directory => {
    const cargo = (hooks.testCandidate ?? testCandidate)(family, directory, candidate);
    assertSources(family, source);
    for (const name of ['family.lock', 'Cargo.lock']) {
      assertIdentity(path.join(family.hub, name), beforeIdentities[name], name);
    }
    for (const repo of family.components()) clean(family.path(repo));
    assertSources(family, source);
    const afterBytes = { 'Cargo.lock': cargo, 'family.lock': candidate };
    recordAfterImages(tx, afterBytes);
    commitPairedReplace(tx, afterBytes, Object.fromEntries(
      Object.entries(beforeIdentities).map(([name, id]) => [name, identityMeta(id)]),
    ), { observe: hooks.observe });
  });
  console.log('family pull: validated candidate locks ready for a protected PR');
}

export function update(family, hooks = {}, tx) {
  if (!tx) return operation(family.hub, inner => updateWithTx(family, hooks, inner), { observe: hooks.observe });
  return updateWithTx(family, hooks, tx);
}

function testCandidate(family, directory, candidate) {
  atomicWrite(path.join(directory, 'candidate.lock'), candidate);
  // Strip credentials and Git rewrites before any candidate checkout/bootstrap.
  const script = fileURLToPath(new URL('./check-family-candidate.mjs', import.meta.url));
  run([process.execPath, script, family.hub, gitText(family.hub, 'rev-parse', 'HEAD'), directory], { env: buildEnvironment() });
  return fs.readFileSync(path.join(directory, 'jankurai/Cargo.lock'), 'utf8');
}
